### PR TITLE
Fuzzing of encoding/decoding functions

### PR DESCRIPTION
**Pull request by xhanulik**
This PR adds fuzzers for encoding functions for PKCS#15 objects and generic `sc_*()` functions.
* `fuzz_pkcs15_encode` iterates over objects on `p15card` and calls corresponding `sc_pkcs15_encode_*_entry()`.
* `fuzz_card` calls some functions from `libopensc/sec.c` and `libopensc/card.c` that are not covered according to OSS-Fuzz coverage reports.

Also, there are some changes in the `fuzz_pkcs15_decode`. The setting of algorithms for the `sc_pkcs15_decode_pubkey()` in a cycle extends testing to other algorithms besides RSA. Some of the `sc_pkcs15_decode_*_entry()` internally use `p15card->app` - these paths were not covered, since `p15card` was not initialized via `sc_pkcs15_bind()`. That is fixed with the usage of the reader from `fuzz_pkcs15_reader`.

After local run, `fuzz_card` encounter asserts in `muscle.c` - these are exchanged for explicit checking the arguments.

##### Checklist

- [x] New files have a LGPL 2.1 license statement

### CODE REVIEW

File: src/tests/fuzzing/fuzz_pkcs15_decode.c
Comment (Jakuje) [inline src/tests/fuzzing/fuzz_pkcs15_decode.c:29]: This part deserves a comment why it is needed now.

File: src/tests/fuzzing/fuzz_pkcs15_decode.c
Comment (Jakuje) [inline src/tests/fuzzing/fuzz_pkcs15_decode.c:107]: I think this should use a new `buf` and `buf_len` gathered from `fuzz_get_chunk()`. This is aliased with the data parsed by `sc_pkcs15_parse_tokeninfo()`.

File: src/tests/fuzzing/fuzz_pkcs15_decode.c
Comment (xhanulik) [reply src/tests/fuzzing/fuzz_pkcs15_decode.c:30]: Yes, I believe it should be in the fuzzers that use some memory allocation (with this flag, the allocator will return 0 on out-of-memory, according to [SanitizerCommonFlags](https://github.com/google/sanitizers/wiki/SanitizerCommonFlags)) - I already added it into the remaining two with some describing comments.

File: src/tests/fuzzing/fuzz_pkcs15_decode.c
Comment (xhanulik) [reply src/tests/fuzzing/fuzz_pkcs15_decode.c:109]: Indeed, thank you.

File: src/tests/fuzzing/fuzz_pkcs15_decode.c
Comment (Jakuje) [reply src/tests/fuzzing/fuzz_pkcs15_decode.c:30]: And do we need the `verbosity` one? It should be default. The other should be called `allocator_may_return_null` (without the `m`), but I am afraid this will cause lot of failures during the allocations before we will get to a useful code, even though it might increase code coverate with the error code paths.

I would probably keep this out, at least from this PR and at least for now.

Comment (xhanulik) [reply src/tests/fuzzing/fuzz_pkcs15_decode.c]: Ok, understood, I removed it.

File: src/tests/fuzzing/fuzz_pkcs15_encode.c
Comment (xhanulik) [reply src/tests/fuzzing/fuzz_pkcs15_encode.c]: Yes, it's more convenient, it's moved into `fuzzer_reader.c`.

File: src/tests/fuzzing/fuzz_card.c
Comment (Jakuje) [inline src/tests/fuzzing/fuzz_card.c:58]: I think this `data` and `size` should be moved by the size of the already read components on lines 50 and 51 so it is not aliased with the card data.

File: src/tests/fuzzing/fuzz_pkcs15_encode.c
Comment (Jakuje) [inline src/tests/fuzzing/fuzz_pkcs15_encode.c:56]:
```suggestion
		switch (obj->type & SC_PKCS15_TYPE_CLASS_MASK) {
```

**Review by Jakuje — APPROVED** [review]

**Comment by Jakuje** [comment]
@frankmorgner can you check this PR if there is something you would like to do differently before I will merge it (after resolving the conflict indeed)?

File: src/tests/fuzzing/fuzz_pkcs15_decode.c
Comment (frankmorgner) [inline src/tests/fuzzing/fuzz_pkcs15_decode.c:99]: For the encoding/decoding operations that are independent from each other, it would maybe be better to either use separate fuzzing tests for them or to *not* advance the buffer with the fuzzing input (`buf`/`buf_len` before `sc_pkcs15_parse_tokeninfo`, `sc_pkcs15_parse_unusedspace`, for example). This allows a wider range of fuzzing input to be useful for actual testing...

File: src/tests/fuzzing/fuzz_pkcs15_encode.c
Comment (frankmorgner) [inline src/tests/fuzzing/fuzz_pkcs15_encode.c:80]: You do pkcs15 bind and then you are using the parsed object to encode this to a buffer, right?

File: src/tests/fuzzing/fuzzer_reader.c
Comment (frankmorgner) [inline src/tests/fuzzing/fuzzer_reader.c:199]: I think the last time, we excluded the formatting changes until https://github.com/OpenSC/OpenSC/pull/2017 is merged, but I'm fine either way, with or without changing this here.

**Review by frankmorgner — CHANGES_REQUESTED** [review]
I like the ideas for integrating the encoding/decoding functions with the fuzzing input, good job!

Some minor things could be improved, though - please see the inline comments.

File: src/tests/fuzzing/fuzz_pkcs15_encode.c
Comment (xhanulik) [reply src/tests/fuzzing/fuzz_pkcs15_encode.c:80]: Yes, after the card initialization it just iterates over objects stored in `p15card` and tries to encode it (with corresponding function).

File: src/tests/fuzzing/fuzzer_reader.c
Comment (xhanulik) [reply src/tests/fuzzing/fuzzer_reader.c:199]: I moved the formatting changes into separate commit, so can be removed.

Comment (xhanulik) [reply src/tests/fuzzing/fuzzer_reader.c]: Indeed, sorry, removed.

**Comment by Jakuje** [comment]
> I like the ideas for integrating the encoding/decoding functions with the fuzzing input, good job!
> 
> Some minor things could be improved, though - please see the inline comments.

@frankmorgner can you check this PR if there is something to fix from your point of view or are we ready to merge this?

**Pull request merged**
